add tutorials for cross encoder models on Amazon Bedrock (#3278)
opensearch-project/ml-commons: docs/tutorials/rerank/rerank_pipeline_with_Cohere_Rerank_model_on_Amazon_Bedrock.md

    body=body
)
results = json.loads(response.get('body').read())["results"]
print(json.dumps(sorted(results, key=lambda x: x['index']),indent=2))
```

The reranking results are as follows:

```
[
  {
    "index": 0,
    "relevance_score": 0.32418242
  },
  {
    "index": 1,
    "relevance_score": 0.07456104
  },
  {
    "index": 2,
    "relevance_score": 0.7190094
  },
  {
    "index": 3,
    "relevance_score": 0.06124987
  }
]
```

## 1. Create a connector and register the model

To create a connector for the model, send the following request. If you are using self-managed OpenSearch, supply your AWS credentials:
```json
POST /_plugins/_ml/connectors/_create
{
  "name": "Amazon Bedrock cross-encoder model",
  "description": "Test connector for Amazon Bedrock cross-encoder model",
  "version": 1,
  "protocol": "aws_sigv4",
  "credential": {
    "access_key": "your_access_key",
    "secret_key": "your_secret_key",
    "session_token": "your_session_token"
  },
  "parameters": {
    "region": "your_bedrock_model_region_like_us-west-2",
    "service_name": "bedrock",
    "api_version": 2
  },
  "actions": [
    {
      "action_type": "predict",
      "method": "POST",

>       "url": "https://bedrock-runtime.${parameters.region}.amazonaws.com/model/cohere.rerank-v3-5:0/invoke",

Reviewer: Similar feedback like before

Author: fixed on 9482b472b2bc68bba61267c87f484967362e2c5b

    ]
  }
}
```

Each item in the array comprises a query_text and a text_docs string, separated by a .

Alternatively, you can test the model as follows:
```json
POST _plugins/_ml/_predict/text_similarity/your_model_id
{
  "query_text": "What is the capital city of America?",
  "text_docs": [
    "Carson City is the capital city of the American state of Nevada.",
    "The Commonwealth of the Northern Mariana Islands is a group of islands in the Pacific Ocean. Its capital is Saipan.",
    "Washington, D.C. (also known as simply Washington or D.C., and officially as the District of Columbia) is the capital of the United States. It is a federal district.",
    "Capital punishment (the death penalty) has existed in the United States since beforethe United States was a country. As of 2017, capital punishment is legal in 30 of the 50 states."
  ]
}
```

The connector `pre_process_function` transforms the input into the format required by parameters shown previously.

By default, Amazon Bedrock Rerank API output has the following format:
```json

> [

Reviewer 2: This default response seems not correct. From my test, I see
```
[{
        "index": 2,
        "relevance_score": 0.7190094
    },
    {
        "index": 0,
        "relevance_score": 0.32418242
    },
    {
        "index": 1,
        "relevance_score": 0.07456104
    },
    {
        "index": 3,
        "relevance_score": 0.06124987
    }
]
```

})

response = bedrock_runtime_client.invoke_model(
    modelId=modelId,
    contentType=contentType,
    accept=accept, 
    body=body
)
results = json.loads(response.get('body').read())["results"]
print(json.dumps(results, indent=2))
```

The reranking result is ordering by the highest score first:
```
[
  {
    "index": 2,
    "relevance_score": 0.7190094
  },
  {
    "index": 0,
    "relevance_score": 0.32418242
  },
  {
    "index": 1,
    "relevance_score": 0.07456104
  },
  {
    "index": 3,
    "relevance_score": 0.06124987
  }
]
```

You can sort the result by index number.

```python
print(json.dumps(sorted(results, key=lambda x: x['index']),indent=2))
```

The results are as follows:
```
[
  {
    "index": 0,
    "relevance_score": 0.32418242
  },
  {
    "index": 1,
    "relevance_score": 0.07456104
  },
  {
    "index": 2,
    "relevance_score": 0.7190094
  },
  {
    "index": 3,
    "relevance_score": 0.06124987
  }
]
```

## 1. Create a connector and register the model

To create a connector for the model, send the following request. If you are using self-managed OpenSearch, supply your AWS credentials:
```json
POST /_plugins/_ml/connectors/_create
{

>   "name": "Amazon Bedrock cross-encoder model",

Reviewer 2: ```suggestion
  "name": "Amazon Bedrock Cohere rerank model",
```

>   "description": "Test connector for Amazon Bedrock cross-encoder model",

Reviewer 2: ```suggestion
  "description": "Test connector for Amazon Bedrock Cohere rerank model",
```

  "version": 1,
  "protocol": "aws_sigv4",
  "credential": {
    "access_key": "your_access_key",
    "secret_key": "your_secret_key",
    "session_token": "your_session_token"
  },
  "parameters": {
    "service_name": "bedrock",

>     "service_code": "bedrock-runtime",

Reviewer 2: ```suggestion
    "endpoint": "bedrock-runtime",
```

>     "region": "your_bedrock_model_region_like_us",

Reviewer 2: ```suggestion
    "region": "your_bedrock_model_region_like_us-west-2",
```

    "model_name": "cohere.rerank-v3-5:0",
    "api_version": 2
  },
  "actions": [
    {
      "action_type": "PREDICT",
      "method": "POST",

>       "url": "https://${parameters.service_code}.${parameters.region}.amazonaws.com/model/${parameters.model_name}/invoke",

Reviewer 2: ```suggestion
      "url": "https://${parameters. endpoint}.${parameters.region}.amazonaws.com/model/${parameters.model_name}/invoke",
```

        }
        def outputs = params.results;
        def relevance_scores = new Double[outputs.length];
        for (int i=0; i<outputs.length; i++) {
          def index = new BigDecimal(outputs[i].index.toString()).intValue();
          relevance_scores[index] = outputs[i].relevance_score;
        }
        def resultBuilder = new StringBuilder('[');
        for (int i=0; i<relevance_scores.length; i++) {
          resultBuilder.append(' {"name": "similarity", "data_type": "FLOAT32", "shape": [1],');
          resultBuilder.append('"data": [');
          resultBuilder.append(relevance_scores[i]);
          resultBuilder.append(']}');
          if (i<outputs.length - 1) {
            resultBuilder.append(',');
          }
        }
        resultBuilder.append(']');
        return resultBuilder.toString();
      """
    }
  ]
}
```

If using the Amazon Opensearch Service, you can provide an IAM role ARN that allows access to the Amazon Bedrock service. For more information, see [AWS documentation](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ml-amazon-connector.html):
```json
POST /_plugins/_ml/connectors/_create
{

>   "name": "Amazon Bedrock cross-encoder model",

Reviewer 2: ```suggestion
  "name": "Amazon Bedrock Cohere rerank model",
```

>   "description": "Test connector for Amazon Bedrock cross-encoder model",

Reviewer 2: ```suggestion
  "description": "Test connector for Amazon Bedrock Cohere rerank model",
```

  "version": 1,
  "protocol": "aws_sigv4",
  "credential": {
    "roleArn": "your_role_arn_which_allows_access_to_bedrock_model"
  },
  "parameters": {
    "service_name": "bedrock",

>     "service_code": "bedrock-runtime",

Reviewer 2: ```suggestion
    "endpoint": "bedrock-runtime",
```

>     "region": "your_bedrock_model_region_like_us",

Reviewer 2: ```suggestion
    "region": "your_bedrock_model_region_like_us-west-2",
```

    "model_name": "cohere.rerank-v3-5:0",
    "api_version": 2
},
  "actions": [
    {
      "action_type": "PREDICT",
      "method": "POST",

>       "url": "https://${parameters.service_code}.${parameters.region}.amazonaws.com/model/${parameters.model_name}/invoke",

Reviewer 2: ```suggestion
      "url": "https://${parameters. endpoint}.${parameters.region}.amazonaws.com/model/${parameters.model_name}/invoke",
```

        }
        """,
      "post_process_function": """
        if (params.results == null || params.results.length == 0) {
          throw new IllegalArgumentException("Post process function input is empty.");
        }
        def outputs = params.results;
        def relevance_scores = new Double[outputs.length];
        for (int i=0; i<outputs.length; i++) {
          def index = new BigDecimal(outputs[i].index.toString()).intValue();
          relevance_scores[index] = outputs[i].relevance_score;
        }
        def resultBuilder = new StringBuilder('[');
        for (int i=0; i<relevance_scores.length; i++) {
          resultBuilder.append(' {"name": "similarity", "data_type": "FLOAT32", "shape": [1],');
          resultBuilder.append('"data": [');
          resultBuilder.append(relevance_scores[i]);
          resultBuilder.append(']}');
          if (i<outputs.length - 1) {
            resultBuilder.append(',');
          }
        }
        resultBuilder.append(']');
        return resultBuilder.toString();
      """
    }
  ]
}
```

Use the connector ID from the response to register and deploy the model:
```json
POST /_plugins/_ml/models/_register?deploy=true
{

>     "name": "Amazon Bedrock Cross-Encoder model",

Reviewer 2: ```suggestion
    "name": "Amazon Bedrock Cohere rerank model",
```

    "function_name": "remote",
    "description": "test rerank model",
    "connector_id": "your_connector_id"
}
```
Note the model ID in the response; you'll use it in the following steps.

Test the model by using the Predict API:
```json
POST _plugins/_ml/models/your_model_id/_predict
{
  "parameters": {
    "query": "What is the capital city of America?",

>     "texts": [

Reviewer 2: ```suggestion
    "documents": [
```

2. The `similarity` outputs are provided in the order of the input documents; the first similarity result pertains to the first document.

## 2. Reranking pipeline
### 2.1 Ingest test data
```json
POST _bulk
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Carson City is the capital city of the American state of Nevada." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "The Commonwealth of the Northern Mariana Islands is a group of islands in the Pacific Ocean. Its capital is Saipan." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Washington, D.C. (also known as simply Washington or D.C., and officially as the District of Columbia) is the capital of the United States. It is a federal district." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Capital punishment (the death penalty) has existed in the United States since beforethe United States was a country. As of 2017, capital punishment is legal in 30 of the 50 states." }

```
### 2.2 Create a reranking pipeline
```json
PUT /_search/pipeline/rerank_pipeline_bedrock
{

>     "description": "Pipeline for reranking with Bedrock cross-encoder model",

Reviewer 2: ```suggestion
    "description": "Pipeline for reranking with Bedrock Cohere rerank model",
```

2. The `similarity` outputs are provided in the order of the input documents; the first similarity result pertains to the first document.

## 2. Reranking pipeline
### 2.1 Ingest test data
```json
POST _bulk
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Carson City is the capital city of the American state of Nevada." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "The Commonwealth of the Northern Mariana Islands is a group of islands in the Pacific Ocean. Its capital is Saipan." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Washington, D.C. (also known as simply Washington or D.C., and officially as the District of Columbia) is the capital of the United States. It is a federal district." }
{ "index": { "_index": "my-test-data" } }
{ "passage_text" : "Capital punishment (the death penalty) has existed in the United States since beforethe United States was a country. As of 2017, capital punishment is legal in 30 of the 50 states." }

```
### 2.2 Create a reranking pipeline
```json

> PUT /_search/pipeline/rerank_pipeline_sagemaker

Reviewer: Same here for the name and description

Author: Thanks, fixed at https://github.com/opensearch-project/ml-commons/pull/3278/commits/a219218a1cbf5930bcc871be7d7016fc7058bb03

            "Its <strong>capital</strong> <strong>is</strong> Saipan."
          ]
        }
      },
      {
        "_index": "my-test-data",
        "_id": "4",
        "_score": 0.06124987,
        "fields": {
          "passage_text": [
            "Capital punishment (the death penalty) has existed in the United States since beforethe United States was a country. As of 2017, capital punishment is legal in 30 of the 50 states."
          ]
        },
        "highlight": {
          "passage_text": [
            "<strong>Capital</strong> punishment (<strong>the</strong> death penalty) has existed in <strong>the</strong> United States since beforethe United States",
            "As <strong>of</strong> 2017, <strong>capital</strong> punishment <strong>is</strong> legal in 30 <strong>of</strong> <strong>the</strong> 50 states."
          ]
        }
      }
    ]
  },
  "profile": {
    "shards": []
  }
}
```

Note: You can avoid writing the query twice by using the `query_text_path` instead of `query_text` as follows:
```json

> POST my-test-data/_search?search_pipeline=rerank_pipeline_sagemaker

Reviewer: Make sure to change it where you write SageMaker too

Author: Thanks, fixed at https://github.com/opensearch-project/ml-commons/pull/3278/commits/a219218a1cbf5930bcc871be7d7016fc7058bb03
